text/204-golang-bindings.md

```diff
@@ -0,0 +1,762 @@
+}
+```
+
+Go interface:
+
+```go
+type ISecurityGroup interface {
+    IResource
+    IPeer
+
+    SecurityGroupId() string
+    AllowAllOutbound() bool
+    AddIngressRule(peer: IPeer, connect: Port, description: *string, remoteRule: *boolean)
+    AddEgressRule(peer: IPeer, connect: Port, description: *string, remoteRule: *boolean)
+}
+```
+
+### JSII Datatype Interfaces (Structs)
+
+In jsii, the InterfaceType has a [datatype field](https://github.com/aws/jsii/blob/master/packages/%40jsii/spec/lib/assembly.ts#L879-L888) attribute
+that indicates that the interface only contains readonly properties. While this does corresponds directly to a Go struct, we would likely need to
+generate both a Go interface that contains getter methods that correspond to each property as well as a Go struct that implements that interface. This
+is in order to support subtyping, as the interface is typically what is passed as an argument into other functions, as well as to ensure forward
+compatibility in case the datatype interface eventually extends another. Were it not for these considerations, it would be simpler to simply have a
+single Go struct that corresponds to a datatype interface (see last bullet point in Notes/Concerns).
+
+#### Case 1: Typescript struct/datatype interface (no extensions)
+
+In the case of a simple a datatype interface (`HealthCheck`) without extensions, we would generate both a Go struct (HealthCheck), which would hold
+the properties of the original typescript interface as public members, and a Go interface (`IHealthCheck`) that defines getter methods for each
+property that would be implemented by the corresponding struct (jsii datatype properties always are readonly, so there are no setter methods
+generated). Since Go does not allow a struct to have exported members with the same name as an interface method, we would have to prefix the interface
+methods; the recommendation here is to use a `Get` prefix.
+
+[Example](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-ecs/lib/container-definition.ts#L596):
+
+```ts
+export interface HealthCheck {
+  readonly command: string[];
+  readonly interval?: cdk.Duration;
+  readonly retries?: number;
+  readonly startPeriod?: cdk.Duration;
+  readonly timeout?: cdk.Duration;
+}
+
+function renderHealthCheck(hc: HealthCheck): cfntaskdefinition.HealthCheckProperty {
+  return {
+    command: getHealthCheckCommand(hc),
+    interval: hc.interval != null ? hc.interval.toSeconds() : 30,
+    retries: hc.retries !== undefined ? hc.retries : 3,
+    startPeriod: hc.startPeriod && hc.startPeriod.toSeconds(),
+    timeout: hc.timeout !== undefined ? hc.timeout.toSeconds() : 5,
+  };
+}
+```
+
+Go translation (Run example in the [Go playground](https://play.golang.org/p/o84uoUGORYH):
+
+```go
+package ecs
+
+type IHealthCheck interface{
+   GetCommand() []string
+   GetInterval() cdk.Duration;
+   GetRetries() int
+   GetStartPeriod() cdk.Duration;
+   GetTimeout() cdk.Duration;
+}
```

Comment:
This example is missing using pointer for the TS optional members. It may also be helpful to callout how optional properties, (and I assume function parameters) are translated into Go.